samza-core/src/main/java/org/apache/samza/application/ApplicationDescriptorImpl.java

```diff
@@ -118,12 +125,21 @@ public S withMetricsReporterFactories(Map<String, MetricsReporterFactory> report
   }
 
   /**
-   * Get the {@link ContextManager} associated with this application
-   *
-   * @return the {@link ContextManager} for this application
+   * @return {@link ApplicationContainerContextFactory} if application specified it; empty otherwise
+   */
+  public Optional<ApplicationContainerContextFactory<ApplicationContainerContext>> getApplicationContainerContextFactory() {
+    //noinspection unchecked; ok because all context types are at least ApplicationContainerContext
+    return Optional.ofNullable(
```

Comment:
Using Optional as a field type has always been something I've been unsure of. I like Optional as a field type since it forces the empty check, but a couple points against it is that it is still possible to set it to null and the original intention was for it to just be for return types.
Since you mention it though, I'm good with using it as a field type.

Reply:
The "noinspection unchecked" acts like an annotation. I see that the annotation is used more throughout the code, so I'll change to that.